add ansible_name/ansible_aliases attrs to plugin objects

s-hertel opened the pull request:
##### SUMMARY
Fix listing filters/tests with --metadata-dump and --list by correcting the plugin names.

Marked a draft until tests are added.

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### COMPONENT NAME
plugin loader

s-hertel committed:
WIP - add new _fqcn attribute to plugin objects

ci_complete

s-hertel committed:
unbreak plugins in subdirs

ci_complete

s-hertel committed:
Fix inadventent changes to _load_name and use existing vars

ci_complete

felixfontein commented:
This PR (rebased to current `devel`) fixes devel ansible-docs for me (by no longer screwing up FQCNs and mass-producing warnings such as `ansible.builtin.urn has a documentation formatting error: 'utf-8' codec can't decode byte 0xf3 in position 1: invalid continuation byte`). It does however produce a strange `ansible.builtin. has a documentation formatting error: 'utf-8' codec can't decode byte 0xf3 in position 1: invalid continuation byte` warning.

s-hertel committed:
add plugin aliases and name property, and replace plugin._load_name where incorrect

fix unit test

s-hertel committed:
Fix listing plugin names

Fix listing legacy and builtin together

test deprecated plugin documentation

fix doc extensions

remove sometimes inaccurate _load_name handling from plugin.name

s-hertel commented:
@webknjaz the bot will just remove the label again until ci is done...

bcoca commented on lib/ansible/plugins/__init__.py:
```diff
@@ -59,6 +59,36 @@ def __init__(self):
+            raise AnsibleError("No plugin name found")
+
+    @property
+    def aliases(self):
```
i would cache it and only construct on empty cache

bcoca commented on lib/ansible/plugins/__init__.py:
```diff
@@ -59,6 +59,36 @@ def __init__(self):
+                names.append(self._fqcn)
+        if self._redirected_names:
+            # FIXME: the first may not be fqcn but not in ansible.builtin (due to collections keyword)
+            names.extend(self._redirected_names[::-1])
```
why not add 'all redirected names'?

s-hertel replied:
there are two `:`, it's just reversing the order to best to worst

bcoca commented on lib/ansible/plugins/loader.py:
```diff
@@ -819,12 +822,13 @@ def _load_module_source(self, name, path):
         setattr(obj, '_original_path', path)
         setattr(obj, '_load_name', name)
         setattr(obj, '_redirected_names', redirected_names or [])
+        setattr(obj, '_fqcn', resolved)
```
tempted to just set the 'name' and aliases at this point

s-hertel replied:
yeah, that seems like a better idea

bcoca replied:
need to x2 check, but probably shoudl happen when 'class_only' also (probably except filters/tests)

s-hertel replied:
I kept it for object instantiation too. The name/aliases from the instantiated object are used in lib/ansible/executor/task_executor.py, lib/ansible/plugins/action/gather_facts.py, and lib/ansible/vars/plugins.py.

If any plugins already have a `name`/`aliases` attribute, that's probably an issue. So I probably need to rename to something less generic... `ansible_internal_name`/`ansible_internal_aliases`?

s-hertel committed:
Add tests for REJECT_EXTS and doc extensions

Fix unpredictable collection redirects so non-fqcns in the redirect list are guaranteed to be legacy (instead of determined by the collections keyword)

Move aliases and name properties to _update_object so all plugin types, including doc fragments, can use them

bcoca commented on lib/ansible/constants.py:
```diff
@@ -100,7 +100,7 @@ def __getitem__(self, y):
 COLLECTION_PTYPE_COMPAT = {'module': 'modules'}
 
-PYTHON_DOC_EXTENSIONS = ('.py', '.pyc', '.pyo')
+PYTHON_DOC_EXTENSIONS = ('.py',)
```
any reason we want to remove compiled versions?

s-hertel replied:
`read_docstub()` was trying to read .pyc files which caused warnings like `[WARNING]: ansible.builtin.urlsplit has a documentation formatting error: 'utf-8' codec can't decode byte 0xf3 in position 1: invalid continuation byte`. @felixfontein caught the bug by having a `ansible/lib/ansible/plugins/test/core.pyc` file and it prevented listing any builtin filters/tests.

bcoca replied:
i wonder if that ever worked

s-hertel replied:
I don't *think* so. But it only became an issue once we started loading adjacent documentation.

s-hertel committed:
fix quotes

s-hertel committed:
remove straggler

s-hertel committed:
make legacy plugin names internally consistent

s-hertel committed:
rename attributes to ansible_name and ansible_aliases

s-hertel committed:
unbreak listing legacy plugins and add a test

s-hertel committed:
remove change to redirects_list

s-hertel commented:
This is not really complete without https://github.com/ansible/ansible/pull/78755 because depending on the collection redirects, the redirect list could contain non-fqcn names that do not correlate to ansible.legacy, and this could cause false positives for `plugin.matches_name()`. But I think this is otherwise ready to merge, and 78755 can be merged before or after.

s-hertel reviewed (commented):
oh, is missing changelog still

s-hertel committed:
changelog

s-hertel commented:
@felixfontein I think this is ready to merge, but I'm going to be out tomorrow/wednesday. I'm not going to rush this in before people have a chance to review, but it will be at the top of my agenda again thursday.

bcoca commented on lib/ansible/plugins/list.py:
```diff
@@ -116,7 +123,6 @@ def list_collection_plugins(ptype, collections, search_paths=None):
 
     # starts at  {plugin_name: filepath, ...}, but changes at the end
     plugins = {}
-    dirs = []
```
keep as if collections is empty, dirs won't be defined for 146/152

s-hertel replied:
It's defined in the `if`, `elif`, and `else` and then used in `_list_plugins_from_paths` for each collection in the loop, so it is always defined. This was to prevent searching irrelevant dirs multiple times when there are multiple collections.

It does look like there's a bug listing all collections when collections is empty (because list_collections isn't getting the playbook adj search path). I will fix that and add a test (tomorrow).

Edit: Actually, there may be nothing to fix there. `ansible-doc` seems to have a requirement on Galaxy files (galaxy.yml or MANIFEST.json) to list arbitrary collections. I just didn't have one of those files in any of my collections.

bcoca commented on lib/ansible/plugins/list.py:
```diff
@@ -126,10 +132,10 @@ def list_collection_plugins(ptype, collections, search_paths=None):
         if collection == 'ansible.builtin':
             # dirs from ansible install, but not configured paths
-            dirs.extend([d.path for d in ploader._get_paths_with_context() if d.internal])
+            dirs = [d.path for d in ploader._get_paths_with_context() if d.internal]
```
with multiple collections this would overwrite previous ones, not add to them

s-hertel replied:
Listing `ansible.legacy` before this change would search the legacy paths twice - once for ansible.legacy, and again for ansible.builtin. This caused issues for differentiating between `ansible.builtin` and `ansible.legacy` (which is probably partly why the `do_replace_legacy` logic existed). So I separated the search paths per loop - now each collection searches only its own paths for plugins.

felixfontein commented:
@s-hertel @bcoca thanks a lot for doing and reviewing this!
